dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/QueryItemHelper.java

```diff
@@ -278,24 +283,84 @@ private static List<Option> getItemOptionsForEmptyRows( EventQueryParams params
      */
     private static List<Option> getItemOptionsThatMatchesRows( Grid grid, int columnIndex )
     {
-        List<Option> options = new ArrayList<>();
         GridHeader gridHeader = grid.getHeaders().get( columnIndex );
 
-        options.addAll( gridHeader
+        return gridHeader
             .getOptionSetObject()
             .getOptions()
             .stream()
             .filter( opt -> opt != null && grid.getRows().stream().anyMatch( r -> {
                 Object o = r.get( columnIndex );
-                if ( o instanceof String )
-                {
-                    return ((String) o).equalsIgnoreCase( opt.getCode() );
-                }
 
-                return false;
-            } ) ).collect( toList() ) );
+                return isItemOptionEqualToRowContent( opt.getCode(), o );
+            } ) ).collect( Collectors.toList() );
+    }
 
-        return options;
+    /**
+     * Compare option code and row content. Type of option value is derived from
+     * ValueType enum
+     *
+     * @see org.hisp.dhis.common.ValueType
+     * @param code
+     * @param rowContent
+     * @return true when equal
+     */
+    private static boolean isItemOptionEqualToRowContent( String code, Object rowContent )
+    {
+        if ( StringUtils.isBlank( code ) )
+        {
+            return false;
+        }
+
+        // String
+        if ( rowContent instanceof String )
+        {
+            return ((String) rowContent).equalsIgnoreCase( code );
+        }
+
+        // Integer, Double
+        if ( rowContent instanceof Number && isNumeric( code ) )
+        {
+            try
+            {
+                return ((Number) rowContent).doubleValue() == Double.parseDouble( code );
+            }
+            catch ( NumberFormatException ignored )
+            {
+            }
```

Comment:
would it make sense to add at least a log here?

Reply:
done, warning